CI: Spark tests fail on TF 1.15 trying to parse "invalid" Keras version '2.2.4-tf'

```
________________________ SparkKerasTests.test_fit_model ________________________
self = <test_spark_keras.SparkKerasTests testMethod=test_fit_model>

    def test_fit_model(self):
        model = create_xor_model()
>       optimizer = get_sgd_optimizer()

test_spark_keras.py:98: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test_spark_keras.py:83: in get_sgd_optimizer
    if version.parse(tf.keras.__version__) < version.parse("2.11"):
/usr/local/lib/python3.7/dist-packages/packaging/version.py:52: in parse
    return Version(version)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <[AttributeError("'Version' object has no attribute '_version'") raised in repr()] Version object at 0x7fdfd7528090>
version = '2.2.4-tf'

    def __init__(self, version: str) -> None:
        """Initialize a Version object.
    
        :param version:
            The string representation of a version which will be parsed and normalized
            before use.
        :raises InvalidVersion:
            If the ``version`` does not conform to PEP 440 in any way then this
            exception will be raised.
        """
    
        # Validate the version and parse it into pieces
        match = self._regex.search(version)
        if not match:
>           raise InvalidVersion(f"Invalid version: '{version}'")
E           packaging.version.InvalidVersion: Invalid version: '2.2.4-tf'

/usr/local/lib/python3.7/dist-packages/packaging/version.py:197: InvalidVersion
```

Seen for `Build and Test (test-cpu-gloo-py3_7-tf1_15_5-keras2_2_4-torch1_8_1-mxnet1_5_1_p0-pyspark3_3_1)`: https://github.com/horovod/horovod/actions/runs/3647365170/jobs/6159518835

Presumably this is related to https://github.com/pypa/packaging/pull/407 in `packaging` version 22.0, released recently: https://github.com/pypa/packaging/releases/tag/22.0

--- a/horovod/_keras/__init__.py
+++ b/horovod/_keras/__init__.py
@@ -254,7 +254,7 @@ def reducescatter(backend, value, name, op):
 
 
 def load_model(keras, wrap_optimizer, optimizer_modules, filepath, custom_optimizers, custom_objects):
-    if version.parse(keras.__version__) < version.parse("2.11"):
+    if version.parse(keras.__version__.replace("-tf", "+tf")) < version.parse("2.11"):
         keras_subclasses = keras.optimizers.Optimizer.__subclasses__()
     else:
         keras_subclasses = keras.optimizers.legacy.Optimizer.__subclasses__()

--- a/horovod/keras/__init__.py
+++ b/horovod/keras/__init__.py
@@ -282,7 +282,7 @@ def load_model(filepath, custom_optimizers=None, custom_objects=None, compressio
     """
     def wrap_optimizer(cls):
         return lambda **kwargs: DistributedOptimizer(cls(**kwargs), compression=compression)
-    if version.parse(keras.__version__) < version.parse("2.11"):
+    if version.parse(keras.__version__.replace("-tf", "+tf")) < version.parse("2.11"):
         optimizer_modules = {keras.optimizers.Optimizer.__module__}
     else:
         optimizer_modules = {keras.optimizers.legacy.Optimizer.__module__}

--- a/horovod/spark/keras/estimator.py
+++ b/horovod/spark/keras/estimator.py
@@ -230,7 +230,7 @@ class KerasEstimator(HorovodEstimator, KerasEstimatorParamsReadable,
             if isinstance(optimizer, str):
                 pass
             else:
-                if version.parse(tf.keras.__version__) < version.parse("2.11"):
+                if version.parse(tf.keras.__version__.replace("-tf", "+tf")) < version.parse("2.11"):
                     if not isinstance(optimizer, tf.keras.optimizers.Optimizer):
                         raise ValueError(f"optimizer has to be an instance of tensorflow.keras.optimizers.Optimizer before Keras 2.11: {type(optimizer).__name__}")
                 else:

--- a/horovod/spark/keras/optimizer.py
+++ b/horovod/spark/keras/optimizer.py
@@ -25,7 +25,7 @@ def serialize_bare_keras_optimizer(x):
     import keras
     from horovod.spark.keras.bare import save_bare_keras_optimizer
 
-    if version.parse(keras.__version__) < version.parse("2.11"):
+    if version.parse(keras.__version__.replace("-tf", "+tf")) < version.parse("2.11"):
         optimizer_class = keras.optimizers.Optimizer
     else:
         optimizer_class = keras.optimizers.legacy.Optimizer
@@ -45,7 +45,7 @@ def serialize_tf_keras_optimizer(x):
     import tensorflow as tf
     from horovod.spark.keras.tensorflow import save_tf_keras_optimizer
 
-    if version.parse(tf.keras.__version__) < version.parse("2.11"):
+    if version.parse(tf.keras.__version__.replace("-tf", "+tf")) < version.parse("2.11"):
         optimizer_class = tf.keras.optimizers.Optimizer
     else:
         optimizer_class = tf.keras.optimizers.legacy.Optimizer

--- a/horovod/tensorflow/keras/__init__.py
+++ b/horovod/tensorflow/keras/__init__.py
@@ -49,7 +49,7 @@ from horovod.tensorflow.keras import callbacks, elastic
 try:
     # In later versions of TensorFlow, optimizers are spread across multiple modules. This set is used to distinguish
     # stock optimizers that come with tf.keras from custom optimizers that may need to be wrapped specially.
-    if version.parse(keras.__version__) < version.parse("2.11"):
+    if version.parse(keras.__version__.replace("-tf", "+tf")) < version.parse("2.11"):
         optimizer_type = tf.keras.optimizers.Optimizer
     else:
         optimizer_type = keras.optimizers.legacy.Optimizer

--- a/test/integration/test_spark_keras.py
+++ b/test/integration/test_spark_keras.py
@@ -80,7 +80,7 @@ def get_mock_fit_fn():
 
 
 def get_sgd_optimizer():
-    if version.parse(tf.keras.__version__) < version.parse("2.11"):
+    if version.parse(tf.keras.__version__.replace("-tf", "+tf")) < version.parse("2.11"):
         optimizer = tf.keras.optimizers.SGD(lr=0.1)
     else:
         optimizer = tf.keras.optimizers.legacy.SGD(lr=0.1)
@@ -212,7 +212,7 @@ class SparkKerasTests(tf.test.TestCase):
 
     def test_fit_model_multiclass(self):
         model = create_mnist_model()
-        if version.parse(tf.keras.__version__) < version.parse("2.11"):
+        if version.parse(tf.keras.__version__.replace("-tf", "+tf")) < version.parse("2.11"):
             optimizer = tf.keras.optimizers.Adadelta(1.0)
         else:
             optimizer = tf.keras.optimizers.legacy.Adadelta(1.0)

--- a/test/parallel/test_tensorflow2_keras.py
+++ b/test/parallel/test_tensorflow2_keras.py
@@ -29,7 +29,7 @@ from tensorflow import keras
 from horovod.common.util import is_version_greater_equal_than
 
 if is_version_greater_equal_than(tf.__version__, "2.6.0"):
-    if version.parse(keras.__version__) < version.parse("2.9.0"):
+    if version.parse(keras.__version__.replace("-tf", "+tf")) < version.parse("2.9.0"):
         from keras.optimizer_v2 import optimizer_v2
     else:
         from keras.optimizers.optimizer_v2 import optimizer_v2
@@ -64,7 +64,7 @@ class Tf2KerasTests(tf.test.TestCase):
 
     def test_train_model_lr_schedule(self):
         initial_lr = 0.1 * hvd.size()
-        if version.parse(tf.keras.__version__) < version.parse("2.11"):
+        if version.parse(tf.keras.__version__.replace("-tf", "+tf")) < version.parse("2.11"):
             opt = tf.keras.optimizers.Adam()
         else:
             opt = tf.keras.optimizers.legacy.Adam()
@@ -157,7 +157,7 @@ class Tf2KerasTests(tf.test.TestCase):
 
     def test_sparse_as_dense_with_grad_aggregation(self):
         backward_passes_per_step = 2
-        if version.parse(keras.__version__) < version.parse("2.11"):
+        if version.parse(keras.__version__.replace("-tf", "+tf")) < version.parse("2.11"):
             opt = keras.optimizers.RMSprop(lr=0.0001)
         else:
             opt = keras.optimizers.legacy.RMSprop(lr=0.0001)
@@ -186,7 +186,7 @@ class Tf2KerasTests(tf.test.TestCase):
     def test_grad_aggregation_with_inf_grad(self):
         backward_passes_per_step = 2
         step_count = tf.Variable(0, trainable=False, dtype=tf.int32)
-        if version.parse(tf.keras.__version__) < version.parse("2.11"):
+        if version.parse(tf.keras.__version__.replace("-tf", "+tf")) < version.parse("2.11"):
             opt = tf.keras.optimizers.SGD()
         else:
             opt = tf.keras.optimizers.legacy.SGD()
@@ -214,7 +214,7 @@ class Tf2KerasTests(tf.test.TestCase):
         assert tf.math.is_finite(grads_and_vars[0][0])
 
     def test_from_config(self):
-        if version.parse(keras.__version__) < version.parse("2.11"):
+        if version.parse(keras.__version__.replace("-tf", "+tf")) < version.parse("2.11"):
             opt = keras.optimizers.Adam()
         else:
             opt = keras.optimizers.legacy.Adam()
@@ -536,7 +536,7 @@ class Tf2KerasTests(tf.test.TestCase):
             model.add(tf.keras.layers.Dense(2, input_shape=(3,), kernel_initializer=initializer, bias_initializer=initializer))
             model.add(tf.keras.layers.RepeatVector(3))
             model.add(tf.keras.layers.TimeDistributed(tf.keras.layers.Dense(3, kernel_initializer=initializer, bias_initializer=initializer)))
-            if version.parse(tf.keras.__version__) < version.parse("2.11"):
+            if version.parse(tf.keras.__version__.replace("-tf", "+tf")) < version.parse("2.11"):
                 opt = tf.keras.optimizers.Adam()
             else:
                 opt = tf.keras.optimizers.legacy.Adam()

--- a/test/parallel/test_tensorflow2_keras_process_sets.py
+++ b/test/parallel/test_tensorflow2_keras_process_sets.py
@@ -60,7 +60,7 @@ class Tf2KerasProcessSetsTests(tf.test.TestCase):
         if size == 1:
             self.skipTest("Only one worker available")
 
-        if version.parse(keras.__version__) < version.parse("2.11"):
+        if version.parse(keras.__version__.replace("-tf", "+tf")) < version.parse("2.11"):
             optimizer_class = keras.optimizers.Optimizer
         else:
             optimizer_class = keras.optimizers.legacy.Optimizer

--- a/test/parallel/test_tensorflow_keras.py
+++ b/test/parallel/test_tensorflow_keras.py
@@ -30,7 +30,7 @@ from horovod.common.util  import is_version_greater_equal_than
 
 if is_version_greater_equal_than(tf.__version__, "2.6.0"):
     from keras import backend as K
-    if version.parse(keras.__version__) < version.parse("2.9.0"):
+    if version.parse(keras.__version__.replace("-tf", "+tf")) < version.parse("2.9.0"):
         from keras.optimizer_v2 import optimizer_v2
     else:
         from keras.optimizers.optimizer_v2 import optimizer_v2

